hsds/util/chunkUtil.py

```diff
@@ -1148,63 +1272,132 @@
+        if where_field not in field_names:
+            msg = f"where field {where_field} is not a member of dataset type"
+            raise ValueError(msg)
+        where_elements = _getWhereElements(query)
+        if not where_elements:
+            msg = "query: where key word with no elements"
+            raise ValueError(msg)
+        # convert to ndarray, checking that we can convert to our dtype along the way
+        try:
+            where_elements_arr = np.array(where_elements, dtype=dset_dt[where_field])
+        except ValueError:
+            msg = "where elements are not compatible with field datatype"
+            raise ValueError(msg)
+        isin_mask = np.isin(chunk_sel[where_field], where_elements_arr)
+
+        if not np.any(isin_mask):
+            # all false
+            log.debug("query - no rows found for where elements")
+            return None
+
+        isin_indices = np.where(isin_mask)
+        if not isinstance(isin_indices, tuple):
+            log.warn(f"expected where_indices of tuple but got: {type(isin_indices)}")
+            return None
+        if len(isin_indices) == 0:
+            log.warn("chunkQuery - got empty tuple where in result")
+            return None
+
+        isin_indices = isin_indices[0]
+        if not isinstance(isin_indices, np.ndarray):
+            log.warn(f"expected isin_indices of ndarray but got: {type(isin_indices)}")
+            return None
+        nrows = isin_indices.shape[0]
+    elif eval_str:
+        log.debug("no where keyword")
+        isin_indices = None
+    else:
+        log.warn("query  - no eval and no where in, returning None")
+        return None
+
     if query_update:
-        replace_mask = [
-            None,
-        ] * len(field_names)
+        if where_field:
+            msg = "query update is not supported with where in"
+            raise ValueError(msg)
+        replace_mask = [None,] * len(field_names)
         for i in range(len(field_names)):
             field_name = field_names[i]
             if field_name in query_update:
                 replace_mask[i] = query_update[field_name]
         log.debug(f"chunkQuery - replace_mask: {replace_mask}")
-        replace_fields = [
-            None,
-        ] * len(field_names)
+        replace_fields = [None, ] * len(field_names)
         if replace_mask == replace_fields:
             msg = "chunkQuery - no fields found in query_update"
             raise ValueError(msg)
     else:
         replace_mask = None
 
-    x = chunk_arr[slices]
-    eval_str = _getEvalStr(query, "x", field_names)
-    where_indices = np.where(eval(eval_str))
-    if not isinstance(where_indices, tuple):
-        log.warn(f"expected where_indices of tuple but got: {type(where_indices)}")
-        return None
-    if len(where_indices) == 0:
-        log.warn("chunkQuery - got empty tuple where result")
-        return None
+    if eval_str:
+        where_indices = np.where(eval(eval_str))
```

Comment:
This evaluation string comes from the user's query, so we need to check to make sure it can't be an injection attack. Similar to the bucket name validation, I think restricting the user's query to only contain alphanumeric characters, spaces, underscores, `>`, `<`, `=`, `"`, and `'` should suffice.